### PR TITLE
Binance,OKx: Implement fetching funding rates

### PR DESCRIPTION
**Pull request by gloriousCode**
# PR Description
This updates the wrapper functions for getting funding rates and implements them for Binance & OKx

This does have some work stolen from my PR in #1220, but I'll just update the other PR after whichever gets merged first

Here's some examples of it in action:
```
go run . f getfundingrates --exchange="okx" --asset="perpetualswap" --pair="btc-usdt-swap" --includepredicted --respecthistorylimits --start="2023-06-18 18:00:00" --includepayments --paymentcurrency="USDT"
{
 "rates": {
  "exchange": "okx",
  "asset": "perpetualswap",
  "pair": {
   "delimiter": "-",
   "base": "BTC",
   "quote": "USDT-SWAP"
  },
  "payment_currency": "USDT",
  "start_date": "2023-06-18 18:00:00 AEST",
  "end_date": "2023-06-20 15:00:00 AEST",
  "rates": [
   {
    "date": "2023-06-20 10:00:00 AEST",
    "rate": "-0.0000503697157989",
    "payment": "0.1541588624312897"
   },
   {
    "date": "2023-06-20 02:00:00 AEST",
    "rate": "0.0000491017037746",
    "payment": "-0.1480155049536702"
   },
   {
    "date": "2023-06-19 18:00:00 AEST",
    "rate": "0.0000072170415113",
    "payment": "-0.0217456663342897"
   },
   {
    "date": "2023-06-19 10:00:00 AEST",
    "rate": "-0.0000423125854412",
    "payment": "0.127107599041561"
   },
   {
    "date": "2023-06-19 02:00:00 AEST",
    "rate": "0.0001987697944021",
    "payment": "-0.6018386023311421"
   }
  ],
  "latest_rate": {
   "date": "2023-06-20 18:00:00 AEST",
   "rate": "0.0000193826255568"
  },
  "upcoming_rate": {
   "date": "2023-06-21 02:00:00 AEST",
   "rate": "-0.0001180215965676"
  },
  "payment_sum": "-0.4903333121462513",
  "time_of_next_rate": "2023-06-21 02:00:00 AEST"
 }
```

```
go run . f getlatestfundingrate --exchange="binance" --asset="coinmarginedfutures" --pair="btcusd-perp" 
{
 "rate": {
  "exchange": "binance",
  "asset": "coinmarginedfutures",
  "pair": {
   "delimiter": "_",
   "base": "BTCUSD",
   "quote": "PERP"
  },
  "latest_rate": {
   "date": "2023-06-20 10:00:00 AEST",
   "rate": "0.003"
  },
  "time_of_next_rate": "2023-06-20 18:00:00 AEST"
 }

```

Much like in #1220, you may need to use the testnet to actually try these features out. Happy to assist, but the instructions are in the other PR.

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## How has this been tested

- [x] go test ./... -race
- [x] golangci-lint run
- [x] Coverage of all new functionality

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation and regenerated documentation via the documentation tool
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally and on Github Actions/AppVeyor with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: engine/rpcserver.go [hunk 1]
```diff
@@ -4665,7 +4660,7 @@ func (s *RPCServer) GetFundingRates(ctx context.Context, r *gctrpc.GetFundingRat
 	if !a.IsFutures() {
 		return nil, fmt.Errorf("%s %w", a, order.ErrNotFuturesAsset)
 	}
-	start := time.Now().AddDate(-1, 0, 0)
+	start := time.Now().AddDate(0, -10, 0)
```
Comment (shazbert) [inline engine/rpcserver.go:4663]: ME: ?

File: exchanges/binance/binance_test.go [hunk 2]
```diff
@@ -2815,3 +2804,135 @@ func TestUpdateOrderExecutionLimits(t *testing.T) {
...
+	}
+
+	is, err = b.IsPerpetualFutureCurrency(asset.USDTMarginedFutures, currency.NewPair(currency.BTC, currency.USDT))
+	if err != nil {
+		t.Error(err)
+	}
+	if !is {
+		t.Error("expected true")
+	}
+	is, err = b.IsPerpetualFutureCurrency(asset.USDTMarginedFutures, currency.NewPair(currency.BTC, currency.PERP))
```
Comment (shazbert) [inline exchanges/binance/binance_test.go:2922]: This looks like a perpy perp to me. Should this be tested for USDT margined?

File: exchanges/binance/binance_types.go [hunk 3]
```diff
@@ -917,3 +918,22 @@ type update struct {
...
+// UserMarginInterestHistoryResponse user margin interest history response
+type UserMarginInterestHistoryResponse struct {
+	Rows  []UserMarginInterestHistory `json:"rows"`
+	Total int64                       `json:"total"`
+}
+
+// UserMarginInterestHistory user margin interest history row
+type UserMarginInterestHistory struct {
+	TxID                int64       `json:"txId"`
+	InterestAccruedTime binanceTime `json:"interestAccuredTime"` // typo in docs, cannot verify due to API restrictions
```
Comment (shazbert) [inline exchanges/binance/binance_types.go:931]: 🤣

File: exchanges/fundingrate/fundingrate_types.go [hunk 4]
```diff
@@ -0,0 +1,70 @@
+package fundingrate
```
Comment (shazbert) [inline exchanges/fundingrate/fundingrate_types.go:1]: could rn package to funding to reduce stutter

File: exchanges/okx/okx.go [hunk 5]
```diff
@@ -3435,8 +3435,8 @@ func (ok *Okx) GetOpenInterest(ctx context.Context, instType, uly, instID string
 	return resp, ok.SendHTTPRequest(ctx, exchange.RestSpot, getOpenInterestEPL, http.MethodGet, common.EncodeURLValues(publicOpenInterestValues, params), nil, &resp, false)
 }
 
-// GetFundingRate  Retrieve funding rate.
-func (ok *Okx) GetFundingRate(ctx context.Context, instrumentID string) (*FundingRateResponse, error) {
+// GetSingleFundingRate  Retrieve funding rate.
```
Comment (shazbert) [inline exchanges/okx/okx.go:3438]: xtra space between GetSingleFundingRate & Retrieve

File: exchanges/okx/okx_types.go [hunk 6]
```diff
@@ -63,6 +64,11 @@ const (
 	operationLogin       = "login"
 )
 
+// testNetKey this key is designed for using the testnet endpoints
+// setting context.WithValue(ctx, testNetKey("testnet"), useTestNet)
+// will ensure the appropriate headers are sent to OKx to use the testnet
+type testNetKey string
```
Comment (shazbert) [inline exchanges/okx/okx_types.go:70]: could also declare a variable below and use that.

**Review by shazbert — CHANGES_REQUESTED** [review]
Looking good!

File: engine/rpcserver.go (hunk 1 above)
Comment (gloriousCode) [reply engine/rpcserver.go:4663]: -10 months seems strange and accidental. Will change to -1 month

File: exchanges/binance/binance_test.go (hunk 2 above)
Comment (gloriousCode) [reply exchanges/binance/binance_test.go:2922]: Binance has no pairs with quote "perp" for perpetual usdtmarginedfutures. I will add a valid usdtmarginedfutures scenario though

File: exchanges/fundingrate/fundingrate_types.go (hunk 4 above)
Comment (gloriousCode) [reply exchanges/fundingrate/fundingrate_types.go:1]: I think funding is too vague for it to be solely about funding rates. funding package for backtester is about managing and maintaining funds for a run as a counter-example

File: cmd/gctcli/futures.go [hunk 7]
```diff
@@ -228,34 +228,73 @@ var futuresCommands = &cli.Command{
...
+				&cli.BoolFlag{
+					Name:    "respecthistorylimits",
+					Aliases: []string{"respect", "r"},
+					Usage:   "optional - if true, will change the starting date to the maximum allowable limit if start date exceeds it",
+				},
+			},
+		},
+		{
+			Name:      "getlatestfundingrate",
+			Aliases:   []string{"latestrate", "lr", "r8"},
```
Comment (thrasher-) [inline cmd/gctcli/futures.go:274]: gr8 r8s m8

File: cmd/gctcli/futures.go [hunk 8]
```diff
@@ -228,34 +228,73 @@ var futuresCommands = &cli.Command{
...
 				&cli.StringFlag{
 					Name:        "end",
 					Aliases:     []string{"ed"},
-					Usage:       "<end> rounded down to the nearest hour, ensure your last position is within this window for accurate calculations",
-					Value:       time.Now().Format(time.DateTime),
+					Usage:       "<end> rounded down to the nearest hour",
+					Value:       time.Now().Truncate(time.Hour).Format(time.DateTime),
 					Destination: &endTime,
 				},
+				&cli.StringFlag{
```
Comment (thrasher-) [inline cmd/gctcli/futures.go:250]: Needs to be added to the ArgsUsage blurb

File: cmd/gctcli/futures.go [hunk 9]
```diff
@@ -228,34 +228,73 @@ var futuresCommands = &cli.Command{
...
-					Usage:   "include the predicted next funding rate",
+					Usage:   "optional - include the predicted next funding rate",
 				},
 				&cli.BoolFlag{
 					Name:    "includepayments",
 					Aliases: []string{"pay"},
-					Usage:   "include funding rate payments",
+					Usage:   "optional - include funding rate payments, must be authenticated",
+				},
+				&cli.BoolFlag{
```
Comment (thrasher-) [inline cmd/gctcli/futures.go:265]: Same with this one

File: cmd/gctcli/futures.go [hunk 10]
```diff
@@ -228,34 +228,73 @@ var futuresCommands = &cli.Command{
...
+					Aliases: []string{"respect", "r"},
+					Usage:   "optional - if true, will change the starting date to the maximum allowable limit if start date exceeds it",
+				},
+			},
+		},
+		{
+			Name:      "getlatestfundingrate",
+			Aliases:   []string{"latestrate", "lr", "r8"},
+			Usage:     "returns the latest funding rate data",
+			ArgsUsage: "<exchange> <asset> <pairs> <start> <end> <includepredicted> <includepayments>",
```
Comment (thrasher-) [inline cmd/gctcli/futures.go:276]:
```suggestion
			ArgsUsage: "<exchange> <asset> <pairs> <includepredicted> <includepayments>",
```

File: cmd/gctcli/futures.go [hunk 11]
```diff
@@ -758,6 +793,13 @@ func getFundingRates(c *cli.Context) error {
 			endTime = c.Args().Get(4)
 		}
 	}
+
+	if c.IsSet("paymentcurrency") {
+		paymentCurrency = c.String("paymentcurrency")
+	} else {
+		paymentCurrency = c.Args().Get(3)
```
Comment (thrasher-) [inline cmd/gctcli/futures.go:800]: This index needs to be adjusted to 5 and the others bumped by one below

File: cmd/gctcli/futures.go [hunk 12]
```diff
@@ -774,6 +816,15 @@ func getFundingRates(c *cli.Context) error {
 			return err
 		}
 	}
+	if c.IsSet("respecthistorylimits") {
+		respectFundingRateHistoryLimits = c.Bool("respecthistorylimits")
+	} else if c.Args().Get(7) != "" {
+		respectFundingRateHistoryLimits, err = strconv.ParseBool(c.Args().Get(6))
```
Comment (thrasher-) [inline cmd/gctcli/futures.go:822]:
```suggestion
		respectFundingRateHistoryLimits, err = strconv.ParseBool(c.Args().Get(7))
```

But will need to bumped by one anyway as per the above comment

File: common/convert/convert.go [hunk 13]
```diff
@@ -226,3 +226,8 @@ func (f *StringToFloat64) UnmarshalJSON(data []byte) error {
 func (f *StringToFloat64) Float64() float64 {
 	return float64(*f)
 }
+
+// Decimal returns the decimal value of the FloatString.
```
Comment (thrasher-) [inline common/convert/convert.go:230]: Please add basic test coverage

File: gctscript/wrappers/validator/validator_test.go [hunk 14]
```diff
@@ -155,7 +155,7 @@ func TestWrapper_Pairs(t *testing.T) {
 
 	_, err = testWrapper.Pairs(exchError.String(), false, asset.Spot)
 	if err == nil {
-		t.Fatal("expected Pairs to return error on invalid name")
+		t.Fatal("expected Pair to return error on invalid name")
```
Comment (thrasher-) [inline gctscript/wrappers/validator/validator_test.go:158]: Can revert as it's valid in this context :)

File: exchanges/okx/okx_type_convert.go [hunk 15]
```diff
@@ -160,6 +160,22 @@ func (a *OpenInterest) UnmarshalJSON(data []byte) error {
...
+	}{
+		Alias: (*Alias)(a),
+	}
+	err := json.Unmarshal(data, chil)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
```
Comment (shazbert) [inline exchanges/okx/okx_type_convert.go]: Rm as this is purging the rates when using test `TestGetFundingRates` also please add field `RealisedRate` in `FundingRateResponse`

```
{"fundingRate":"0.0001905490218961","fundingTime":"1687536000000","instId":"BTC-USD-SWAP","instType":"SWAP","realizedRate":"0.0001905452237828"}
```

**Review by shazbert — CHANGES_REQUESTED** [review]
Thanks, Just one thing on recent changes.

File: exchanges/okx/okx_type_convert.go (hunk 15 above)
Comment (gloriousCode) [reply exchanges/okx/okx_type_convert.go]: I don't know why its showing up for you that I introduced that, but its a great pickup! I would have been [assigning the incorrect historical rate](https://www.okx.com/docs-v5/en/#public-data-rest-api-get-funding-rate-history) 🦅 🎉

**Review by shazbert — APPROVED** [review]
tACK nice!

**Review by thrasher- — APPROVED** [review]
Thanks for making those changes and tACK, great stuff!

image

getfundingrates:

image

Matching their front-end (if you * 100):

image

**Pull request merged**